File: client/app/src/main/java/com/rh/heji/data/db/Image.java
package com.rh.heji.data.db;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Ignore;
import androidx.room.PrimaryKey;

import java.util.Objects;

/**
 * Date: 2020/11/19
 * Author: 锅得铁
 * #
 */
@Entity(tableName = Image.TAB_NAME, foreignKeys = @ForeignKey(entity = Bill.class,
        parentColumns = "bill_id",
        childColumns = "_bid",
        onDelete = ForeignKey.CASCADE,
        onUpdate = ForeignKey.CASCADE
))
public class Image {

    public static final String TAB_NAME = "bill_img";
    public static final String COLUMN_ID = "_bid";
    public static final String COLUMN_PATH = "img_path";
    public static final String COLUMN_ONLINE_PATH = "img_online_path";
    public static final String COLUMN_STATUS = "sync_status";

    @NonNull
    @PrimaryKey(autoGenerate = true)
    Long _id;


    @NonNull
    @ColumnInfo(name = COLUMN_ID,index = true)
    String bill_id;

    String md5;

    private String ext;

    @ColumnInfo(name = COLUMN_PATH)
    String localPath;

    @ColumnInfo(name = COLUMN_ONLINE_PATH)
    String onlinePath;

    @ColumnInfo(name = COLUMN_STATUS, defaultValue = "0")
    int synced;

    public Image() {
    }

    @Ignore
    public Image(String bill_id) {
        this.bill_id = bill_id;
    }

    public Long getId() {
        return _id;
    }

    public void setId(Long id) {
        this._id = id;
    }

    public String getLocalPath() {
        return localPath;
    }

    public void setLocalPath(String localPath) {
        this.localPath = localPath;
    }

    public String getOnlinePath() {
        return onlinePath;
    }

    public void setOnlinePath(String onlinePath) {
        this.onlinePath = onlinePath;
    }

    public int getSynced() {
        return synced;
    }

    public void setSynced(int synced) {
        this.synced = synced;
    }

    @NonNull
    public String getBillImageID() {
        return bill_id;
    }

    public void setBillImageID(@NonNull String billImageID) {
        this.bill_id = billImageID;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Image image = (Image) o;
        return _id == image._id &&
                bill_id.equals(image.bill_id) &&
                Objects.equals(localPath, image.localPath) &&
                Objects.equals(onlinePath, image.onlinePath);
    }

    public String getMd5() {
        return md5;
    }

    public void setMd5(String md5) {
        this.md5 = md5;
    }

    @Override
    public int hashCode() {
        return Objects.hash(_id, bill_id, localPath, onlinePath);
    }

    public String getExt() {
        return ext;
    }

    public void setExt(String ext) {
        this.ext = ext;
    }

    @Override
    public String toString() {
        return "Image{" +
                "_id='" + _id + '\'' +
                ", bill_id='" + bill_id + '\'' +
                ", md5='" + md5 + '\'' +
                ", ext='" + ext + '\'' +
                ", localPath='" + localPath + '\'' +
                ", onlinePath='" + onlinePath + '\'' +
                ", synced=" + synced +
                '}';
    }
}

File: client/app/src/main/java/com/rh/heji/data/db/Image.kt
package com.rh.heji.data.db

import androidx.room.*
import com.rh.heji.data.db.mongo.ObjectId
import java.util.*

/**
 * Date: 2020/11/19
 * Author: 锅得铁
 * primaryKeys image_id
 * ForeignKey bill_id -> image.bill_id
 */
@Entity(
    tableName = Image.TAB_NAME,
    primaryKeys = [Image.COLUMN_ID],
    foreignKeys = [ForeignKey(
        entity = Bill::class,
        parentColumns = [Bill.COLUMN_ID],
        childColumns = [Image.COLUMN_BILL_ID],
        onDelete = ForeignKey.CASCADE,
        onUpdate = ForeignKey.NO_ACTION
    )], indices = [Index(value = [Image.COLUMN_ID, Image.COLUMN_BILL_ID], unique = true)]
)
data class Image(
    @ColumnInfo(name = COLUMN_ID)
    var id: String= ObjectId().toHexString(),
    @ColumnInfo(name = Bill.COLUMN_ID, index = true)
    var billID: String
) {


    var md5: String? = null
    var ext: String? = null

    @ColumnInfo(name = COLUMN_PATH)
    var localPath: String? = null

    @ColumnInfo(name = COLUMN_ONLINE_PATH)
    var onlinePath: String? = null

    @ColumnInfo(name = COLUMN_STATUS, defaultValue = "0")
    var synced = 0

    override fun hashCode(): Int {
        return Objects.hash(id, billID, localPath, onlinePath)
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val image = o as Image
        return id === image.id && billID == image.billID &&
                localPath == image.localPath &&
                onlinePath == image.onlinePath
    }

    companion object {
        const val COLUMN_ID = "image_id"
        const val COLUMN_BILL_ID = Bill.COLUMN_ID
        const val TAB_NAME = "image"
        const val COLUMN_PATH = "local_path"
        const val COLUMN_ONLINE_PATH = "online_path"
        const val COLUMN_STATUS = "sync_status"
    }

    override fun toString(): String {
        return "Image(id='$id', billID='$billID', md5=$md5, ext=$ext, localPath=$localPath, onlinePath=$onlinePath, synced=$synced)"
    }
}